### PR TITLE
Issue/async bootloader

### PR DESCRIPTION
**Pull request by arnaudsjs**

### CODE REVIEW

**Comment by bartv** [comment]
Ok but I added wouter because he wrote the initial implementation

File: src/inmanta/app.py [hunk 1]
```diff
@@ -55,31 +58,47 @@
...
     ibl = InmantaBootloader()
-    ibl.start()
 
-    try:
-        io_loop.start()
-    except KeyboardInterrupt:
-        IOLoop.current().stop()
-        ibl.stop()
+    def stop_server_signal_handler(signum, frame):
+        IOLoop.current().add_callback_from_signal(stop, ibl.stop)
```
Comment (wouterdb) [inline src/inmanta/app.py]: this code had me confused, could be a little more clear 
either

- give the stop function a clearer name (safe_shutdown_wrapper?)
- move the whole signal handling thing into a method that takes the stop function (ibl.stop I mean, not stop) as an argument and does all the signal setup?

File: src/inmanta/server/protocol.py [hunk 2]
```diff
@@ -161,13 +162,16 @@ def __init__(self, name):
...
+    @gen.coroutine
     def prestart(self, server: RESTServer):
         """Called by the RestServer host prior to start, can be used to collect references to other server slices"""
         pass
 
+    @gen.coroutine
     def start(self):
         pass
 
+    @gen.coroutine
```
Comment (wouterdb) [inline src/inmanta/server/protocol.py]: please document that fact that 
1- stop is blocking
2- there is at this point no way to indicate the startup/shutdown order

File: src/inmanta/server/bootloader.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 from inmanta.server import server
 from inmanta.server.protocol import RESTServer
 from inmanta.server.agentmanager import AgentManager
+from tornado import gen
 
 
 class InmantaBootloader(object):
```
Comment (wouterdb) [inline src/inmanta/server/bootloader.py]: document that stop is blocking and startup/shutdown order is hardcoded.

File: src/inmanta/server/protocol.py [hunk 4]
```diff
@@ -130,16 +131,16 @@ def start(self):
...
         self.http_server.listen(port)
 
         LOGGER.debug("Start REST transport")
 
+    @gen.coroutine
     def stop(self):
-        LOGGER.debug("Stoppin Server Rest Endpoint")
+        LOGGER.debug("Stopping Server Rest Endpoint")
         self.http_server.stop()
         for endpoint in self.__end_points:
```
Comment (wouterdb) [inline src/inmanta/server/protocol.py]: stopping in starting order will be a problem if the database connection is properly closed

should be reverse order imho

File: src/inmanta/server/bootloader.py (hunk 3 above)
Comment (arnaudsjs) [reply src/inmanta/server/bootloader.py]: What do you mean with "stop is blocking"? stop() is an asynchronous function, so it's non-blocking. Right?

Comment (wouterdb) [reply src/inmanta/server/bootloader.py]: I mean that it only returns if the stop is complete (i.e. there is no join)

on second thought, that is pretty default, so no need to document

**Pull request merged**